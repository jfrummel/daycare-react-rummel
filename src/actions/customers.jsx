import { v4 as uuidv4 } from 'uuid';

//  Customers action generators

export const addCustomer = ({ firstName = "", lastName = "", numOfChildren = 0, hourlyRate = 0 } = {}) => ({
    type: "ADD_CUSTOMER",
    customer: {
        id: uuidv4(),
        firstName,
        lastName,
        numOfChildren,
        hourlyRate
    }
});

export const removeCustomer = ({ id } = {}) => ({
    type: "REMOVE_CUSTOMER",
    id
});

export const editCustomer = (id, updates) => ({
    type: "EDIT_CUSTOMER",
    id, updates
});