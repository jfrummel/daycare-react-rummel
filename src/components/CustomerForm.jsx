import React, { Component } from 'react';



class CustomerForm extends Component {
    constructor(props) {
        super(props);

        this.state = {
            firstName: props.customer ? props.customer.firstName : "",
            lastName: props.customer ? props.customer.lastName : "",
            numOfChildren: props.customer ? props.customer.numOfChildren : 0,
            hourlyRate: props.customer ? (props.customer.hourlyRate.toString()) : "",
            error: ""
        }
    }

    onFirstNameChange = (e) => {
        const firstName = e.target.value;
        this.setState(() => ({ firstName }));
    };

    onLastNameChange = (e) => {
        const lastName = e.target.value;
        this.setState(() => ({ lastName }));
    };

    onNumChildrenChange = (e) => {
        const numOfChildren = e.target.value;
        this.setState(() => ({ numOfChildren }));
    };

    onHourlyRateChange = (e) => {
        const hourlyRate = e.target.value;
        if (!hourlyRate || hourlyRate.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ hourlyRate }));
        }
    };

    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.firstName || !this.state.lastName) {
            this.setState(() => ({ error: "First and Last Name Required" }))
        } else {
            this.setState(() => ({ error: "" }));
            this.props.onSubmit({
                firstName: this.state.firstName,
                lastName: this.state.lastName,
                numOfChildren: this.state.numOfChildren,
                hourlyRate: parseFloat(this.state.hourlyRate)
            });
        };
    };

    render() {
        return (
            <form onSubmit={this.onSubmit}>
                {this.state.error && <p>{this.state.error}</p>}
                <input
                    type="text"
                    value={this.state.firstName}
                    placeholder="First Name"
                    onChange={this.onFirstNameChange}
                    autoFocus
                />
                <input
                    type="text"
                    value={this.state.lastName}
                    placeholder="Last Name"
                    onChange={this.onLastNameChange}
                />
                <input
                    type="number"
                    value={this.state.numOfChildren}
                    placeholder="Number of Children"
                    onChange={this.onNumChildrenChange}
                />
                <input
                    type="text"
                    value={this.state.hourlyRate}
                    placeholder="Hourly Rate"
                    onChange={this.onHourlyRateChange}
                />
                <button>{this.props.buttonText}</button>
            </form>
        );
    }
}

export default CustomerForm;