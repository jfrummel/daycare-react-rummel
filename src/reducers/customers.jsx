//  customers default state

const customerDefaultState = [];

const customerReducer = (state = customerDefaultState, action) => {
    switch (action.type) {
        case "ADD_CUSTOMER":
            return [...state, action.customer];
        case "REMOVE_CUSTOMER":
            return state.filter(({ id }) => id !== action.id);
        case "EDIT_CUSTOMER":
            return state.map((customer) => {
                if (customer.id === action.id) {
                    return { ...customer, ...action.updates };
                } else {
                    return customer
                }
            });
        default:
            return state;
    }
};

export default customerReducer;