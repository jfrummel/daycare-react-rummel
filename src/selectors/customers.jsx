
export const getVisibleCustomers = (customers, { text }) => {
    return customers.filter((customer) => {
        const textMatch = customer.lastName.toLowerCase().includes(text.toLowerCase()) ||
            customer.firstName.toLowerCase().includes(text.toLowerCase());

        return textMatch;
    });
};