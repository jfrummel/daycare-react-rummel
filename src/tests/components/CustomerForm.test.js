import React from 'react';
import { shallow } from 'enzyme';
import CustomerForm from '../../components/CustomerForm';
import { customers } from '../fixtures/customers';

test("renders form with default values", () => {
    const wrapper = shallow(<CustomerForm />);
    expect(wrapper).toMatchSnapshot();
});

test("renders form with customer values", () => {
    const wrapper = shallow(<CustomerForm customer={customers[1]} />);
    expect(wrapper).toMatchSnapshot();
});

test("sets error message with invalid form submission", () => {
    const wrapper = shallow(<CustomerForm />);
    wrapper.find('form').simulate('submit', {
        preventDefault: () => { }
    });
    expect(wrapper.state('error').length).toBeGreaterThan(0);
});

test("sets state for first name", () => {
    const value = "test";
    const wrapper = shallow(<CustomerForm />);
    wrapper.find('input').at(0).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('firstName')).toBe(value);
});

test("sets state for last name", () => {
    const value = "test last name";
    const wrapper = shallow(<CustomerForm />);
    wrapper.find('input').at(1).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('lastName')).toBe(value);
});

test("sets state for number of children", () => {
    const value = 7;
    const wrapper = shallow(<CustomerForm />)
    wrapper.find('input').at(2).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('numOfChildren')).toBe(value);
});

test("sets state for hourly rate", () => {
    const value = "1.50";
    const wrapper = shallow(<CustomerForm />);
    wrapper.find('input').at(3).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('hourlyRate')).toBe(value);
});

test("does not set state of hourly rate with invalid data", () => {
    const value = "1.2563";
    const wrapper = shallow(<CustomerForm />);
    wrapper.find('input').at(3).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('hourlyRate')).toBe("");
});

test("calls on submit with correct data", () => {
    const onSubmitSpy = jest.fn();
    const wrapper = shallow(<CustomerForm customer={customers[1]} onSubmit={onSubmitSpy} />)
    wrapper.find('form').simulate('submit', {
        preventDefault: () => { }
    });
    expect(wrapper.state('error')).toBe("");
    expect(onSubmitSpy).toHaveBeenLastCalledWith({
        firstName: customers[1].firstName,
        lastName: customers[1].lastName,
        numOfChildren: customers[1].numOfChildren,
        hourlyRate: customers[1].hourlyRate
    });
});