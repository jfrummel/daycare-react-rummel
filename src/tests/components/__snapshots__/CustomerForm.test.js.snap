// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders form with customer values 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    onChange={[Function]}
    type="text"
    value="Lauren"
  />
  <input
    onChange={[Function]}
    type="text"
    value="Victoria"
  />
  <input
    onChange={[Function]}
    type="number"
    value={7}
  />
  <input
    onChange={[Function]}
    type="text"
    value="3"
  />
  <button />
</form>
`;

exports[`renders form with default values 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    onChange={[Function]}
    type="text"
    value=""
  />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
  <input
    onChange={[Function]}
    type="number"
    value={0}
  />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
  <button />
</form>
`;
